content/blog/2021/06/09/StaticWebsiteGenerationWithJavaMavenAndJBake.adoc

```diff
@@ -0,0 +1,248 @@
+This has many advantages over a Content Management System (CMS) for projects such as ours:
+
+* *Hosting* is cheap. https://pages.github.com/[GitHub pages] even hosts static websites for free.
+* The source files go into Git for *backup* and *history*.
+* The source files are *plain text*:
+** Changes come in as a Pull Request for *review* with CI validation.
+** The sources are open in our IDE, which encourages refactoring them alongside the code.
+This results in *less stale* content.
+
+For many years, Awestruct has served us well.
+But due to lack of activity, it was time to upgrade.
+
+== Why JBake?
+
+Because we're Java programmers.
+
+There are several good static website generators out there, such as Jekyll (Ruby) and Hugo (Go).
+We choose https://jbake.org/[JBake] (Java), because:
+
+. Our website now *builds with Maven* (`mvn generate-resources`).
++
+_No need to install anything. Not even JBake._
+Everyone builds with the same version of JBake, as declared in the `pom.xml`.
++
+And it's fast: even a `mvn clean` build of 150 output pages only takes 20 seconds on my machine.
+
+. It's all *Java underneath*.
++
+Writing conditional expressions is straightforward.
+The APIs (`String.substring()`, ...) are familiar.
+Date formatting (`d MMMM yyyy`) and regular expressions behave as expected.
++
+And most importantly, error messages are clear.
+
+For 8 years, I wrote the website with Awestruct (Ruby).
+But I never took the time to decently learn Ruby, so every change entailed hours of trial and error.
+I couldn't just read the error message and fix it.
+This isn't Ruby's fault. It was because I never took a few days to actually learn Ruby.
+With JBake, I fix errors in a fraction of time: no more trial and error.
+
+== What is JBake?
+
+JBake is a static website generator with many options:
+
+* Build with Maven or Gradle.
++
+We choose https://maven.apache.org/[Maven], because all our repos build with Maven
+(although two https://github.com/kiegroup/optaplanner-quickstarts[OptaPlanner Quickstarts] also build with Gradle because OptaPlanner supports Gradle too).
+
+* Write content in Asciidoc, Markdown or HTML.
++
+We choose https://asciidoc.org/[Asciidoc]
+because it's richer and more reliable than Markdown and all our documentation is written in Asciidoc.
+
+* Create templates with Freemarker, Thymeleaf or Groovy.
++
+We choose https://freemarker.apache.org/[Freemarker]
+because it's a powerful, battle-tested templating engine.
+
+== Tips and tricks
+
+Here are common tasks to build an advanced static website
+and how to implement them in JBake-Freemarker.
+You might even call these _JBake Design Patterns_:
+
+=== Use a macro to render shared content
+
+Almost all our templates show the same latest releases panel:
+
+image::staticWebsiteMacroLatestReleases.png[Latest releases screenshot]
+
+A Freemarker templates is perfect to avoid repeating yourself (DRY):
```

Comment:
```suggestion
A Freemarker templates is perfect for avoiding repeating yourself (DRY):
```

Reply:
Pretty sure that "is perfect to avoid" is valid english

Reply:
See some of the examples here https://www.linguee.com/english-french/translation/it+doesn%27t+have+to+be+perfect.html